brainiak/isfc.py

```diff
@@ -1,226 +1,1104 @@
+           "voxel(s) or ROI(s).")
+    
+    # Compute summary statistic for observed ISCs
+    if summary_statistic == np.mean:
+        observed = np.tanh(np.mean(np.arctanh(iscs), axis=0))[np.newaxis, :]
+    elif summary_statistic == np.median:
+        observed = summary_statistic(iscs, axis=0)[np.newaxis, :]
+    else:
+        raise TypeError("Unrecognized summary_statistic! Use np.median or np.mean.")
+    
+    # Set up an empty list to build our bootstrap distribution
+    distribution = []
+    
+    # Loop through n bootstrap iterations and populate distribution
+    for i in np.arange(n_bootstraps):
+        
+        # Random seed to be deterministically re-randomized at each iteration
+        if isinstance(random_state, np.random.RandomState):
+            prng = random_state
+        else:
+            prng = np.random.RandomState(random_state)
+
+        # Randomly sample subject IDs with replacement
+        subject_sample = sorted(prng.choice(np.arange(n_subjects),
+                                                 size=n_subjects))
+        
+        # Squareform and shuffle rows/columns of pairwise ISC matrix to
+        # to retain correlation structure among ISCs, then get triangle
+        if pairwise:
+            
+            # Loop through voxels
+            isc_sample = []
+            for voxel_iscs in iscs.T:
+
+                # Square the triangle and fill diagonal
+                voxel_iscs = squareform(voxel_iscs)
+                np.fill_diagonal(voxel_iscs, 1)
+
+                # Check that pairwise ISC matrix is square and symmetric
+                assert voxel_iscs.shape[0] == voxel_iscs.shape[1]
+                assert np.allclose(voxel_iscs, voxel_iscs.T)
+
+                # Shuffle square correlation matrix and get triangle
+                voxel_sample = voxel_iscs[subject_sample, :][:, subject_sample]
+                voxel_sample = squareform(voxel_sample, checks=False)
+
+                # Censor off-diagonal 1s for same-subject pairs
+                voxel_sample[voxel_sample == 1.] = np.NaN
+
+                isc_sample.append(voxel_sample)
+                
+            isc_sample = np.column_stack(isc_sample)
+
+        # Get simple bootstrap sample if not pairwise
+        elif not pairwise:
+            isc_sample = iscs[subject_sample, :]
+            
+        # Compute summary statistic for bootstrap ISCs per voxel
+        # (alternatively could construct distrubtion for all voxels
+        # then compute statistics, but larger memory footprint)
+        if summary_statistic == np.mean:
+            distribution.append(np.tanh(np.nanmean(np.arctanh(isc_sample), axis=0)))
+        elif summary_statistic == np.median:
+            distribution.append(np.nanmedian(isc_sample, axis=0))
+                    
+        # Update random state for next iteration
+        random_state = np.random.RandomState(prng.randint(0, 2**32 - 1))
```

Comment:
This `2**32` is re-used several times. I think this should be a named constant (e.g. MAX_RAND_SEED) in this script.